### PR TITLE
Changed all the rejections to use an Error parameter (no typings change)

### PR DESCRIPTION
**Pull request by digeff**
@dlebu @joshuaskelton @meenakb @MSLaguana @MShamiltonia @guillaumejenkins
Do you think that forcing all rejections to be an error would improve our code or make it worse?

### CODE REVIEW

File: src/common/nestedError.ts [hunk 1]
```diff
@@ -1,11 +1,13 @@
...
+    private extras: any;
 
-    constructor(message: string, innerError: any) {
+    constructor(message: string, innerError: any, extras?: any) {
         super(message);
         this.innerError = innerError;
         this.name = innerError.name;
         const innerMessage = innerError.message;
         this.message = innerMessage ? `${message}: ${innerMessage}` : message;
+        this.extras = extras;
```
Comment (MSLaguana) [inline src/common/nestedError.ts]: What do we do with this? There's no way to extract it, and it looks like it isn't included in the stringified version

Comment (digeff) [inline src/common/nestedError.ts]: You are right. I'll add a public getter (at the moment we just use it to store stderr on .exec())

File: src/typings/q/Q.d.ts [hunk 2]
```diff
@@ -306,7 +306,7 @@ declare module Q {
     /**
      * A settable property that will intercept any uncaught errors that would otherwise be thrown in the next tick of the event loop, usually as a result of done. Can be useful for getting the full stack trace of an error in browsers, which is not usually possible with window.onerror.
      */
-    export var onerror: (reason: any) => void;
+    export var onerror: (reason: Error) => void;
```
Comment (MSLaguana) [inline src/typings/q/Q.d.ts]: Is this changed intentionally? I don't think I've ever messed with this before.

**Comment by MSLaguana** [comment]
I think that being rigorous about what we reject is a good idea, because there have been issues in the past where someone would throw a string, and we'd fail to get a `.message` out of it or some such.

File: src/typings/q/Q.d.ts (hunk 2 above)
Comment (digeff) [inline src/typings/q/Q.d.ts]: I just changed all the things that seemed reasonable that they could be an error.

File: src/common/nestedError.ts [hunk 3]
```diff
@@ -1,11 +1,17 @@
...
-        this.innerError = innerError;
+        this._innerError = innerError;
         this.name = innerError.name;
         const innerMessage = innerError.message;
         this.message = innerMessage ? `${message}: ${innerMessage}` : message;
+        this._extras = extras;
+    }
+
+    public get extras(): any {
+        return this.extras;
```
Comment (MSLaguana) [inline src/common/nestedError.ts]: You mean this._extras, otherwise it's an infinite loop.

Comment (digeff) [inline src/common/nestedError.ts]: Fixed!

**Comment by MSLaguana** [comment]
Looks good to me

**Comment by joshuaskelton** [comment]
Looks good to me.

**Pull request merged**